### DISCUSSION TITLE
Can't specify just DNS server for PXE booting

### DISCUSSION THREAD

**Discussion by mrichar1** [siderolabs/talos#9500] [Q&A]
I'm trying to PXE boot talos 1.8, having previously been booting 1.3.1 successfully.

Our servers can't reach the outside, so we need to specify the DNS (and NTP) servers, but want to just use DHCP/defaults otherwise.

I'm using the following on the kernel command-line to set the DNS server and NTP server (same host):

```ip=:::::::10.0.0.254::10.0.0.254```

However it seems to try to create a `bond0 interface which then never gets a network address, rather than using the 'physical' interface:

```
8021q: adding VLAN 0 to HW filter on device bond0
Waiting up to 100 more seconds for network.
<... counts to 0 then loops forever ...>
```

I can explictly set the interface by changing the `ip` param to:

```ip=:::::eno1:dhcp:10.0.0.254::10.0.0.254```

This doesn't then try to generate `bond0` but instead seems to fail to being up the interface properly (no DHCP requests are seen on the DHCP server) and the logs show:

```
bnx2 0000:01:00.0 eno1: renamed from eth0
[talos] controller failed {"component": "controller-runtime", "controller": "network.LinkSpecController", "error": "2 errors occurred:\n\t* error changing flags for \"eno1\": netlink receive: no such file or directory\n\t*}
[talos] time query error with server "10.0.0.254" {"component": "controller-runtime", "controller": "time.SyncController", "error": "dial udp 10.0.0.254:123: connect: network is unreachable"}
```

I have also tried using `net.ifnames=0` and setting `eth0` instead of `eno1` but the result is the same:

```
[talos] controller failed {"component": "controller-runtime", "controller": "network.LinkSpecController", "error": "2 errors occurred:\n\t* error changing flags for \"eth0\": netlink receive: no such file or directory\n\t*}
[talos] time query error with server "10.0.0.254" {"component": "controller-runtime", "controller": "time.SyncController", "error": "dial udp 10.0.0.254:123: connect: network is unreachable"}
```

I can boot this server from PXE into other environments and the networking stack all works as expected for DHCP/DNS/NTP.

Any ideas on how to fix this?

**Comment by mrichar1** [comment]
Some progress...

I discovered the 'new' (since 1.6) Image Factory images were needed since my test server requires `bnx2` firmware. Using that image and setting the interface as `eno1` solves the 'no network' problem.

However more generally I still need to be able to not specify the interface name as this varies between different server hardware types in the cluster.

**Comment by smira** [comment]
`bond0` is not created as a result of `ip=` arg, it's rather Linux kernel is configured to create one instance of each interface type, but that can be ignored.

The `ip=:::::::10.0.0.254::10.0.0.254` should work, if it doesn't please open an issue with the logs.

**Comment by mrichar1** [comment]
Having gone back to test this to submit a bug it does eventually succeed, but there is around 3 minutes of `Waiting...` logs. Digging deeper it seems to be trying each interface serially, and takes a long time to timeout and eventually find a valid one.

Looking at the dracut kernel params documentation, it provides the option:

```
single-dhcp
               Send DHCP on all available interfaces in parallel, as
               opposed to one after another. After the first DHCP
               response is received, stop DHCP on all other interfaces.
               This gives the fastest boot time by using the IP on
               interface for which DHCP succeeded first during early
               boot.
```

I can confirm that setting this in the `ip` line means there is no delay in booting waiting for the network any more. (Presumably this is the default behaviour if no `ip` line is passed, since that is the behaviour seen when not specified).

I'll see about making a documentation PR to add this to the talos kernel docs as this might be generally more useful than `dhcp` if specifying...

**Reply by smira** [reply]
I think you're triggering Linux kernel to do DHCP with `ip=` line, it's not Talos itself.

**Reply by mrichar1** [reply]
yes - sorry I wasn't clear - this definitely isn't a bug - just a side-effect of how dracut does dhcp (compounded by my missing the requirements for an Image Factory kernel with bnx2 firmware) coupled with having a server with lots of interfaces being tried in serial.

My PR comment was just thinking that it could be worth highlighting that `single-dhcp` is potentially a better option in many cases than `dhcp` which is the option currently recommended in the talos kernel docs for the `ip` option here: https://www.talos.dev/v1.8/reference/kernel/

**Reply by smira** [reply]
There is no dracut in Talos at all.

This is the kernel documentation: https://www.kernel.org/doc/Documentation/filesystems/nfs/nfsroot.txt

Talos re-uses some parts of the cmdline `ip=`, but it ignores kernel doing DHCP.

**Reply by mrichar1** [reply]
Apologies - got my wires crossed with reading the kernel docs this week...

I was really just meaning that where the `talos` kernel docs give examples of using `dhcp` in the `ip` option, it could be worth mentioning `single-dhcp` as a better option for bare-metal/pxe builds on hosts with multiple interfaces. But maybe that's clear enough from reading the kernel reference docs.
